File: api/app/schemas/semantic.py
from typing import Annotated, Literal

from pydantic import BaseModel, Field, field_validator, model_validator

Shape = Literal["distribution", "ranking", "timeseries", "comparison", "single"]
Operator = Literal["eq", "in", "gte", "lte", "between", "score_band", "top_n", "bottom_n"]


class QueryFilter(BaseModel):
    dimension: str
    operator: Operator = "eq"
    value: str | int | float | list[str | int | float]


class QuerySort(BaseModel):
    by: str = "value"
    direction: Literal["asc", "desc"] = "desc"


class SemanticQuery(BaseModel):
    measures: Annotated[list[str], Field(min_length=1, max_length=5)]
    dimensions: Annotated[list[str], Field(default_factory=list, max_length=4)]
    filters: Annotated[list[QueryFilter], Field(default_factory=list, max_length=12)]
    shape: Shape
    sort: QuerySort | None = None
    limit: Annotated[int, Field(default=25, ge=1, le=250)]

    @field_validator("measures", "dimensions")
    @classmethod
    def unique_keys(cls, value: list[str]) -> list[str]:
        if len(value) != len(set(value)):
            raise ValueError("keys must be unique")
        return value

    @model_validator(mode="after")
    def shape_contract(self) -> "SemanticQuery":
        if self.shape == "timeseries" and "fy" not in self.dimensions:
            raise ValueError("timeseries queries require the fy dimension")
        if self.shape == "ranking" and "company" not in self.dimensions:
            raise ValueError("ranking queries require the company dimension")
        return self


class LineageRef(BaseModel):
    pin_id: str
    filing_id: str | None = None
    field_key: str | None = None
    source_page: int | None = None


class PolicyNotice(BaseModel):
    code: str
    message: str
    measure: str | None = None


class SemanticResponse(BaseModel):
    data: list[dict[str, object]]
    lineage_refs: dict[str, list[LineageRef]]
    applied_policy: list[PolicyNotice]
    catalog_version: str
    cache_hit: bool = False

File: frontend/src/lib/semantic.ts
import { useQuery } from '@tanstack/react-query'

import { accessToken } from './auth'
import { apiUrl } from './api'

export type Shape = 'distribution' | 'ranking' | 'timeseries' | 'comparison' | 'single'
export interface SemanticFilter { dimension: string; operator: 'eq' | 'in' | 'gte' | 'lte' | 'between' | 'score_band'; value: string | number | Array<string | number> }
export interface SemanticDSL {
  measures: string[]
  dimensions: string[]
  filters: SemanticFilter[]
  shape: Shape
  sort?: { by: string; direction: 'asc' | 'desc' }
  limit?: number
}
export interface PolicyNotice { code: string; message: string; measure?: string }
export interface SemanticResponse {
  data: Array<Record<string, unknown>>
  lineage_refs: Record<string, Array<{ pin_id: string; filing_id?: string; field_key?: string; source_page?: number }>>
  applied_policy: PolicyNotice[]
  catalog_version: string
  cache_hit: boolean
}
export async function runSemanticQuery(dsl: SemanticDSL): Promise<SemanticResponse> {
  const token = accessToken()
  const response = await fetch(`${apiUrl}/api/query`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json', ...(token ? { Authorization: `Bearer ${token}` } : {}) },
    body: JSON.stringify(dsl),
  })
  if (!response.ok) throw new Error(`Query failed (${response.status})`)
  return response.json() as Promise<SemanticResponse>
}

export function useSemanticQuery(dsl: SemanticDSL, enabled = true) {
  return useQuery({
    queryKey: ['semantic', dsl],
    queryFn: () => runSemanticQuery(dsl),
    staleTime: 300_000,
    enabled,
    // Refining a filter holds the previous result on screen at reduced opacity
    // rather than collapsing the layout into a skeleton.
    placeholderData: previous => previous,
  })
}

export function consolidatePolicyNotices(notices: PolicyNotice[] = []): PolicyNotice[] {
  return [...new Map(notices.map(item => [item.code, item])).values()]
}

export function encodeQueryState(dsl: SemanticDSL): string {
  return new URLSearchParams({ q: JSON.stringify(dsl) }).toString()
}

export function decodeQueryState(search: string, fallback: SemanticDSL): SemanticDSL {
  const raw = new URLSearchParams(search).get('q')
  if (!raw) return fallback
  try { return JSON.parse(raw) as SemanticDSL } catch { return fallback }
}

export function replaceQueryState(dsl: SemanticDSL): void {
  window.history.replaceState(null, '', `${window.location.pathname}?${encodeQueryState(dsl)}`)
}
